llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -7437,3 +7437,212 @@ bool CombinerHelper::matchNonNegZext(const MachineOperand &MO,
 
   return false;
 }
+
+bool CombinerHelper::matchPtrAddWithSub(const MachineOperand &MO,
+                                        BuildFnTy &MatchInfo) {
+  GPtrAdd *Inner = cast<GPtrAdd>(MRI.getVRegDef(MO.getReg()));
+  GSub *Sub = cast<GSub>(MRI.getVRegDef(Inner->getOffsetReg()));
+
+  // sub(x, c) -> add(x, -c)
+
+  // one-use check
+  if (!MRI.hasOneNonDBGUse(Sub->getReg(0)))
+    return false;
+
+  // Cannot fail due to pattern.
+  std::optional<APInt> MaybeImm = getIConstantVRegVal(Sub->getRHSReg(), MRI);
+  if (!MaybeImm)
+    return false;
+
+  LLT ConstTy = MRI.getType(Inner->getOffsetReg());
+
+  if (!isConstantLegalOrBeforeLegalizer(ConstTy))
+    return false;
+
+  Register Dst = MO.getReg();
+  LLT DstTy = MRI.getType(Dst);
+
+  MatchInfo = [=](MachineIRBuilder &B) {
```

Comment:
Can you please try to avoid this pattern (using a "build function") ? 

Seems like a good amount of these could just be MIR patterns, or just normal C++ functions
For instance, you don't have to directly negate the constant, a pattern can just emit `G_NEG` and let another rule do it.

Reply:
There is no `G_NEG`. There is a `G_FNEG`. I must do the negation in C++.

Reply:
Do `G_SUB 0, x`  and let a combine do it?
If it has to be C++ then fine, but I'd prefer without the lambdas and instead using a proper helper

Reply:
IIRC, 0 works in the match part, but it is not legal MIR.

Reply:
I don't understand, do you mean it doesn't work in the apply patterns? Because it does
`G_SUB $sub, 0, $x` should be a valid MIR pattern. No reason it doesn't work, if it doesn't, then it's a bug

Reply:
In the match part, I can say a MachineOperand is a constant, e.g., 0. I bet in MIR that I would need to create a G_CONSTANT.

Reply:
The `G_SUB $negation, 0, $const` pattern caused endless loops, but this was probably me. I moved some patterns into the apply.